misc/test_expected_payload_cooker.py

```diff
@@ -0,0 +1,197 @@
+FAILING_TEST_HEADER_PATTERN = re.compile(r"\W*FAIL \[[ 0-9.]+s\] ([ \w::]+)")
+
+# Expected print message: `***expected: <dump>`
+##     println!("***expected: {:?}", expected.dump().unwrap());
+TAG_PATTERN = re.compile(r"\W*\*\*\*expected: \[([ 0-9,]*)\]")
+
+KEY_CANDIDATES = [
+    binascii.unhexlify("b1b52e16c1b46ab133c8bf576e82d26c887f1e9deae1af80043a258c36fcabf3")
+]
+
+
+def decode_expected_raw(raw: bytes) -> dict[str, object]:
+    def attempt_decrypt(raw: bytes) -> bytes | None:
+        for key in KEY_CANDIDATES:
+            key = nacl.secret.SecretBox(key)
+            try:
+                return key.decrypt(raw)
+            except nacl.exceptions.CryptoError:
+                continue
+
+    def attempt_decompression(raw: bytes) -> bytes | None:
+        try:
+            return zstandard.decompress(raw)
+        except ValueError:
+            return None
+
+    def attempt_deserialization(raw: bytes) -> dict[str, object] | None:
+        if raw[0] == 0xFF:
+            try:
+                return msgpack.unpackb(raw[1:])
+            except ValueError:
+                pass
+
+        # First byte is the version
+        if raw[0] != 0x00:
+            return None
+
+        try:
+            decompressed = zstandard.decompress(raw[1:])
+        except ValueError:
+            return None
+
+        try:
+            return msgpack.unpackb(decompressed)
+        except ValueError:
+            return None
+
+    # First attempt: consider the data is not signed nor encrypted
+    deserialized = attempt_deserialization(raw)
+    if deserialized is not None:
+        return deserialized
+
+    # Second attempt: consider the data is only signed
+    raw_without_signature = raw[64:]
+    deserialized = attempt_deserialization(raw_without_signature)
+    if deserialized is not None:
+        return deserialized
+
+    # Last attempt: consider the data is signed and encrypted
+    decrypted = attempt_decrypt(raw)
+    assert decrypted is not None
+    decrypted_without_signature = decrypted[64:]
+    decompressed = attempt_decompression(decrypted_without_signature)
+    assert decompressed is not None
+    deserialized = attempt_deserialization(decompressed)
+    assert deserialized is not None
+    return deserialized
+
+
+def cook_msgpack_type(value):
+    if isinstance(value, bytes):
+        return value.hex()
+
+    if isinstance(value, msgpack.ExtType):
+        match value.code:
+            # DateTime
+            case 1:
+                ts = struct.unpack(">q", value.data)[0]
+                dt = datetime.datetime.fromtimestamp(ts / 1000000).isoformat() + "Z"
+                return f"ext(1, {ts}) i.e. {dt}"
+            # UUID
+            case 2:
+                return f"ext(2, {value.data.hex()})"
+
+            case _:
+                pass
+
+    return value
+
+
+current_test = None
+
+lines = sys.stdin.readlines()
```

Comment:
Maybe put the  `main` logic in the `main` function, so this code doesn't run if this file is imported by mistake.

Reply:
Done